src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -2630,18 +2648,166 @@ CalculateNewConnectionCount(WorkerPool *workerPool)
+			 *
+			 * Note that we are not trying to find the optimal number of connections
+			 * for the remaining tasks here. Our goal is to prevent connection
+			 * establishments that are absolutely unnecessary. In the future, we may
+			 * improve the calculations below to find the optimal number of new
+			 * connections required.
+			 */
+			if (ExecutorOptimizeWithCurrentExecutionStats &&
+				UsingExistingConnectionsCheaperThanEstablishingNewConnections(
+					readyTaskCount, workerPool))
+			{
+				return 0;
+			}
 		}
 	}
+
 	return newConnectionCount;
 }
 
 
+/*
+ * UsingExistingConnectionsCheaperThanEstablishingNewConnections returns true if
+ * using the already established connections takes less time compared to opening
+ * new connections based on the current execution's stats.
+ *
+ * The function returns false if the current execution has not established any connections
+ * or finished any tasks (e.g., no stats to act on).
+ */
+static bool
+UsingExistingConnectionsCheaperThanEstablishingNewConnections(int newConnectionCount,
+															  WorkerPool *workerPool)
+{
+	int activeConnectionCount = workerPool->activeConnectionCount;
+
+	if (workerPool->totalExecutedTasks < 1 || activeConnectionCount < 1)
+	{
+		/*
+		 * The pool has not finished any connection establishment or
+		 * task yet. So, we refrain from optimizing the execution.
+		 */
+		return false;
+	}
+
+	double avgTaskExecutionTime = AvgTaskExecutionTimeApproximation(workerPool);
+	double avgConnectionEstablishmentTime = AvgConnectionEstablishmentTime(workerPool);
+
+	/* we assume that we are halfway through the execution */
+	double remainingTimeForActiveTaskExecutionsToFinish = avgTaskExecutionTime / 2;
+
+	/*
+	 * We use "newConnectionCount" as if it is the task count as
+	 * we are only interested in this iteration of CalculateNewConnectionCount().
+	 */
+	int taskCount = newConnectionCount;
+	double totalTimeToExecuteNewTasks = avgTaskExecutionTime * taskCount;
+
+	double estimatedExecutionTimeForNewTasks =
+		floor(totalTimeToExecuteNewTasks / activeConnectionCount);
+
+	/*
+	 * First finish the already running tasks, and then use the connections
+	 * to execute the new tasks.
+	 */
+	double costOfExecutingTheTasksOverExistingConnections =
+		remainingTimeForActiveTaskExecutionsToFinish +
+		estimatedExecutionTimeForNewTasks;
+
+	/*
+	 * For every task, the executor is supposed to establish one
+	 * connection and then execute the task over the connection.
+	 */
+	double costOfExecutingTheTasksOverNewConnection =
+		(avgTaskExecutionTime + avgConnectionEstablishmentTime);
+
+	return (costOfExecutingTheTasksOverExistingConnections <=
+			costOfExecutingTheTasksOverNewConnection);
+}
+
+
+/*
+ * AvgTaskExecutionTimeApproximation returns the approximation of the average task
+ * execution times on the workerPool.
+ */
+static double
+AvgTaskExecutionTimeApproximation(WorkerPool *workerPool)
+{
+	double totalTaskExecutionTime = workerPool->totalTaskExecutionTime;
```

Comment:
And then make this one a `uint64` too and only convert to `double` during the division on the last line of this function.